File: 383. Ransom Note/canConstruct.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
const canConstruct = (ransomNote, magazine) => {
  const magazineCharCnt = {};
  for (let i = 0; i < magazine.length; i++) {
    if (!magazineCharCnt[magazine[i]]) {
      magazineCharCnt[magazine[i]] = 0;
    }
    magazineCharCnt[magazine[i]]++;
  }

  for (let i = 0; i < ransomNote.length; i++) {
    if (!magazineCharCnt[ransomNote[i]] || magazineCharCnt[ransomNote[i]] === 0) {
      return false;
    }
    magazineCharCnt[ransomNote[i]]--;
  }

  return true;
};


const ransomNote = "a";
const magazine = "b";
console.log('canConstruct: ', canConstruct(ransomNote, magazine));

File: 383. Ransom Note/canConstruct.py
class Solution:
  def canConstruct(self, ransomNote: str, magazine: str) -> bool:
    magazine_char_cnt = {}
    for c in magazine:
      if c not in magazine_char_cnt:
        magazine_char_cnt[c] = 0
      magazine_char_cnt[c] += 1
      
    for i in range(len(ransomNote)):
      if (ransomNote[i] not in magazine_char_cnt) or magazine_char_cnt[ransomNote[i]] == 0:
        return False
      magazine_char_cnt[ransomNote[i]] -= 1
      
    return True


ransomNote = "a"
magazine = "b"
solution = Solution()
print('canConstruct: ', solution.canConstruct(ransomNote, magazine))
